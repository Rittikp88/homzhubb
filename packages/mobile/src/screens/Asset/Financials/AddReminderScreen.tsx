import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { useRoute } from '@react-navigation/native';
import { useNavigation } from '@react-navigation/core';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { LedgerRepository } from '@homzhub/common/src/domain/repositories/LedgerRepository';
import { FinancialSelectors } from '@homzhub/common/src/modules/financials/selectors';
import { icons } from '@homzhub/common/src/assets/icon';
import { theme } from '@homzhub/common/src/styles/theme';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import ConfirmationSheet from '@homzhub/mobile/src/components/molecules/ConfirmationSheet';
import ReminderForm from '@homzhub/common/src/components/organisms/ReminderForm';
import { IAddReminder, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

const AddReminderScreen = (): React.ReactElement => {
  const { goBack, navigate } = useNavigation();
  const { params } = useRoute();
  const { t } = useTranslation();
  const loaders = useSelector(FinancialSelectors.getFinancialLoaders);
  const [isLoading, setLoading] = useState(false);
  const [isSheetVisible, setSheetVisibility] = useState(false);
  const [showDeleteIcon, setShowDeleteIcon] = useState(false);
  const selectedReminderId = useSelector(FinancialSelectors.getCurrentReminderId);

  const param = params as IAddReminder;

  const onPressIcon = (visible: boolean): void => {
    setSheetVisibility(visible);
  };

  const onPressDelete = async (): Promise<void> => {
    try {
      await LedgerRepository.deleteReminderById(selectedReminderId);
      goBack();
      onPressIcon(false);
      AlertHelper.success({ message: t('assetFinancial:reminderDeleteMsg') });
    } catch (e) {
      onPressIcon(false);
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  const handleAddAccount = (): void => {
    navigate(ScreensKeys.AddBankAccount);
  };

  return (
    <>
      <Screen
        backgroundColor={theme.colors.white}
        isLoading={loaders.reminder || isLoading}
        keyboardShouldPersistTaps
        headerProps={{
          title: param?.isEdit ? t('assetFinancial:editReminder') : t('assetFinancial:addReminders'),
          type: 'secondary',
          onIconPress: goBack,
          ...(showDeleteIcon && { iconRight: icons.trash, onIconRightPress: (): void => onPressIcon(true) }),
        }}
      >
        <ReminderForm
          onSubmit={goBack}
          isEdit={param?.isEdit ?? false}
          isFromDues={param?.isFromDues ?? false}
          setLoading={setLoading}
          onAddAccount={handleAddAccount}
          setShowDeleteIcon={setShowDeleteIcon}
        />
      </Screen>
      <ConfirmationSheet
        isVisible={isSheetVisible}
        sheetTitle={t('common:delete')}
        message={t('property:deleteConfirmation', { name: t('assetFinancial:thisReminder') })}
        onCloseSheet={(): void => onPressIcon(false)}
        onPressDelete={onPressDelete}
      />
    </>
  );
};

export default AddReminderScreen;
