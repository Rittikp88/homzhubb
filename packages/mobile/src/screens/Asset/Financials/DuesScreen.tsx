import React, { useCallback } from 'react';
import { StyleSheet } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { useFocusEffect, useNavigation } from '@react-navigation/native';
import { FinancialActions } from '@homzhub/common/src/modules/financials/actions';
import { FinancialSelectors } from '@homzhub/common/src/modules/financials/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import DueList from '@homzhub/mobile/src/components/organisms/DueList';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';

const DuesScreen = (): React.ReactElement => {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const { goBack } = useNavigation();
  const dueItems = useSelector(FinancialSelectors.getDueItems);
  const { dues: dueLoading } = useSelector(FinancialSelectors.getFinancialLoaders);

  useFocusEffect(
    useCallback(() => {
      dispatch(FinancialActions.getDues());
      dispatch(FinancialActions.setCurrentDueId(-1));
    }, [])
  );

  return (
    <Screen
      backgroundColor={theme.colors.white}
      headerProps={{
        title: t('assetDashboard:dues'),
        type: 'secondary',
        onIconPress: goBack,
      }}
      containerStyle={styles.container}
      isLoading={dueLoading}
    >
      <DueList dues={dueItems} />
    </Screen>
  );
};

export default React.memo(DuesScreen);

const styles = StyleSheet.create({
  container: {
    marginTop: 10,
  },
});
