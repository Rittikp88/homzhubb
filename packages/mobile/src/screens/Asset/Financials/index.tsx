import React from 'react';
import { StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { DateUtils } from '@homzhub/common/src/utils/DateUtils';
import { LedgerUtils } from '@homzhub/common/src/utils/LedgerUtils';
import { LedgerService } from '@homzhub/common/src/services/LedgerService';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { FinancialsNavigatorParamList } from '@homzhub/mobile/src/navigation/BottomTabs';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import {
  AnimatedProfileHeader,
  FinancialHeaderContainer,
  PropertyDuesCardContainer,
  StateAwareComponent,
} from '@homzhub/mobile/src/components';
import FinanceOverview from '@homzhub/mobile/src/components/organisms/FinanceOverview';
import TransactionCardsContainer from '@homzhub/mobile/src/components/organisms/TransactionCardsContainer';
import { propertyDues } from '@homzhub/common/src/mocks/FinancialsTabMockData';
import { DataGroupBy, GeneralLedgers, LedgerTypes } from '@homzhub/common/src/domain/models/GeneralLedgers';
import { FinancialRecords, FinancialTransactions } from '@homzhub/common/src/domain/models/FinancialTransactions';

interface IState {
  ledgerData: GeneralLedgers[];
  transactionsData: FinancialRecords[];
  isLoading: boolean;
  scrollEnabled: boolean;
}

type Props = WithTranslation & NavigationScreenProps<FinancialsNavigatorParamList, ScreensKeys.FinancialsLandingScreen>;

export class Financials extends React.PureComponent<Props, IState> {
  private onFocusSubscription: any;

  public state = {
    ledgerData: [],
    transactionsData: [],
    isLoading: false,
    scrollEnabled: true,
  };

  public componentDidMount(): void {
    const { navigation } = this.props;

    this.onFocusSubscription = navigation.addListener('focus', (): void => {
      this.getGeneralLedgersPref().then();
      this.getGeneralLedgers().then();
    });
  }

  public render = (): React.ReactElement => {
    const { isLoading } = this.state;
    return <StateAwareComponent loading={isLoading} renderComponent={this.renderComponent()} />;
  };

  private renderComponent = (): React.ReactElement => {
    const { t } = this.props;
    const { ledgerData, transactionsData, scrollEnabled } = this.state;
    const { currency_symbol, totalDue, details } = propertyDues;

    return (
      <AnimatedProfileHeader isOuterScrollEnabled={scrollEnabled} title={t('financial')}>
        <>
          <FinancialHeaderContainer
            onPlusIconClicked={this.onPlusIconPress}
            // @ts-ignore
            title={t('assetFinancial:recordsText')}
            income={LedgerUtils.getSumOfTransactionsOfType(LedgerTypes.credit, ledgerData)}
            expense={LedgerUtils.getSumOfTransactionsOfType(LedgerTypes.debit, ledgerData)}
            individualCardStyle={styles.individualCardStyle}
          />
          <FinanceOverview />
          <PropertyDuesCardContainer currency={currency_symbol} totalDue={totalDue} propertyDues={details} />
          {transactionsData && transactionsData.length > 0 && (
            <TransactionCardsContainer
              shouldEnableOuterScroll={this.toggleScroll}
              transactionsData={transactionsData}
              onEndReachedHandler={this.onEndReachedHandler}
            />
          )}
        </>
      </AnimatedProfileHeader>
    );
  };

  private onEndReachedHandler = async (): Promise<void> => {
    await this.getGeneralLedgers();
  };

  private onPlusIconPress = (): void => {
    const {
      navigation: { navigate },
    } = this.props;
    navigate(ScreensKeys.AddRecordScreen);
  };

  private toggleScroll = (scrollEnabled: boolean): void => {
    this.setState({ scrollEnabled });
  };

  private getGeneralLedgersPref = async (): Promise<void> => {
    this.setState({ isLoading: true });
    try {
      const response: GeneralLedgers[] = await LedgerService.getLedgerPerformances(
        DateUtils.getCurrentYearStartDate(),
        DateUtils.getCurrentYearLastDate(),
        DataGroupBy.year
      );

      this.setState({ ledgerData: response, isLoading: false });
    } catch (e) {
      this.setState({ isLoading: false });
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  private getGeneralLedgers = async (): Promise<void> => {
    const { transactionsData } = this.state;
    this.setState(() => {
      return { isLoading: false };
    });

    try {
      const response: FinancialTransactions = await LedgerService.getGeneralLedgers(transactionsData.length);

      this.setState({
        transactionsData: [...transactionsData, ...response.results],
        isLoading: false,
      });
    } catch (e) {
      this.setState({ isLoading: false });
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };
}

export default withTranslation(LocaleConstants.namespacesKey.assetFinancial)(Financials);

const styles = StyleSheet.create({
  individualCardStyle: {
    alignItems: 'center',
    paddingHorizontal: 25,
  },
});
