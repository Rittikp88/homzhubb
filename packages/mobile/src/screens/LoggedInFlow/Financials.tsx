import React from 'react';
import { View, StyleSheet, Text } from 'react-native';

export class Financials extends React.PureComponent<{}, {}> {
  public render = (): React.ReactElement => {
    return (
      <View style={styles.screen}>
        <Text>Financials Screen</Text>
      </View>
    );
  };
}

const styles = StyleSheet.create({
  screen: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
