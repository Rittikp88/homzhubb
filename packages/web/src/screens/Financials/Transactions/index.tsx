import React from 'react';
import { StyleSheet, View, ScrollView } from 'react-native';
import { useTranslation } from 'react-i18next';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import Accordian from '@homzhub/web/src/screens/Financials/Transactions/Accordian';

const Transactions = (): React.ReactElement => {
  const { t } = useTranslation();

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Icon name={icons.wallet} size={22} color={theme.colors.darkTint3} />
        <Text type="small" textType="semiBold" style={styles.text}>
          {t('assetDashboard:Transaction')}
        </Text>
      </View>
      <ScrollView>
        <Accordian />
        <Accordian />
        <Accordian />
        <Accordian />
        <Accordian /> <Accordian />
        <Accordian />
        <Accordian />
        <Accordian />
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    marginVertical: 24,
    borderRadius: 4,
    backgroundColor: theme.colors.white,
    padding: 15,
    height: 600,
  },
  amount: {
    marginRight: 10,
    color: theme.colors.error,
  },
  header: {
    flexDirection: 'row',
    margin: 10,
  },
  leftChild: {
    flexDirection: 'row',
  },
  text: { marginLeft: 10, color: theme.colors.darkTint1 },
});
export default Transactions;
