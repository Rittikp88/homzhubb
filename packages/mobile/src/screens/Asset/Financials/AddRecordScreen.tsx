import React, { ReactElement } from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';
import { StyleSheet } from 'react-native';
import { connect } from 'react-redux';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { LedgerRepository } from '@homzhub/common/src/domain/repositories/LedgerRepository';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { FinancialsNavigatorParamList } from '@homzhub/mobile/src/navigation/BottomTabs';
import { AnimatedProfileHeader, HeaderCard, StateAwareComponent } from '@homzhub/mobile/src/components';
import AddRecordForm from '@homzhub/mobile/src/components/organisms/AddRecordForm';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { Currency } from '@homzhub/common/src/domain/models/Currency';
import { LedgerCategory } from '@homzhub/common/src/domain/models/LedgerCategory';
import { IState } from '@homzhub/common/src/modules/interfaces';

interface IScreenState {
  ledgerCategories: LedgerCategory[];
  clearForm: boolean;
  isLoading: boolean;
}

interface IStateToProps {
  currency: Currency;
  assets: Asset[];
}

type libraryProps = WithTranslation & NavigationScreenProps<FinancialsNavigatorParamList, ScreensKeys.AddRecordScreen>;
type IProps = libraryProps & IStateToProps;

export class AddRecordScreen extends React.PureComponent<IProps, IScreenState> {
  public state = {
    ledgerCategories: [],
    isLoading: false,
    clearForm: false,
  };

  public async componentDidMount(): Promise<void> {
    this.setState({ isLoading: true });

    const categories = await LedgerRepository.getLedgerCategories();

    this.setState({ ledgerCategories: categories, isLoading: false });
  }

  public render(): ReactElement {
    const { isLoading } = this.state;

    return <StateAwareComponent loading={isLoading} renderComponent={this.renderComponent()} />;
  }

  private renderComponent = (): ReactElement => {
    const { t } = this.props;
    const { clearForm } = this.state;
    return (
      <AnimatedProfileHeader title={t('financial')}>
        <>
          <HeaderCard
            title={t('addRecords')}
            clear={clearForm}
            subTitle={t('common:clear')}
            renderItem={this.renderAddRecordForm}
            onIconPress={this.goBack}
            onClearPress={this.onClearPress}
          />
        </>
      </AnimatedProfileHeader>
    );
  };

  private renderAddRecordForm = (): ReactElement => {
    const { clearForm, ledgerCategories } = this.state;
    const { currency, assets } = this.props;

    return (
      <AddRecordForm
        properties={assets}
        ledgerCategories={ledgerCategories}
        clear={clearForm}
        defaultCurrency={currency}
        onFormClear={this.onClearPress}
        containerStyles={styles.addFormContainer}
        onSubmitFormSuccess={this.onSubmitFormSuccess}
        shouldLoad={this.toggleLoading}
      />
    );
  };

  private onClearPress = (): void => {
    const { clearForm } = this.state;
    this.setState({ clearForm: !clearForm });
  };

  private onSubmitFormSuccess = (): void => {
    const { t } = this.props;
    AlertHelper.success({ message: t('addedSuccessfullyMessage') });
    this.goBack();
  };

  private toggleLoading = (isLoading: boolean): void => {
    this.setState({ isLoading });
  };

  private goBack = (): void => {
    const { navigation } = this.props;

    if (navigation.canGoBack()) {
      navigation.goBack();
    }
  };
}

const mapStateToProps = (state: IState): IStateToProps => {
  return {
    currency: UserSelector.getCurrency(state),
    assets: UserSelector.getUserAssets(state),
  };
};

export default connect(
  mapStateToProps,
  null
)(withTranslation(LocaleConstants.namespacesKey.assetFinancial)(AddRecordScreen));

const styles = StyleSheet.create({
  addFormContainer: {
    marginTop: 24,
  },
});
