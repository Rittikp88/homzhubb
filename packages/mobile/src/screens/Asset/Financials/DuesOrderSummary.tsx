import React, { useState, useEffect } from 'react';
import { View, FlatList, StyleSheet } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation } from '@react-navigation/native';
import { useTranslation } from 'react-i18next';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { LedgerRepository } from '@homzhub/common/src/domain/repositories/LedgerRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { FinancialActions } from '@homzhub/common/src/modules/financials/actions';
import { FinancialSelectors } from '@homzhub/common/src/modules/financials/selectors';
import { Divider } from '@homzhub/common/src/components/atoms/Divider';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { Loader } from '@homzhub/common/src/components/atoms/Loader';
import HomzhubCoins from '@homzhub/common/src/components/molecules/HomzhubCoins';
import OrderSummary from '@homzhub/common/src/components/molecules/OrderSummary';
import { PropertyAddressCountry } from '@homzhub/common/src/components/molecules/PropertyAddressCountry';
import PromoCode from '@homzhub/common/src/components/molecules/PromoCode';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { PaymentGateway } from '@homzhub/mobile/src/components';
import { Coins } from '@homzhub/common/src/domain/models/OrderSummary';
import { DueOrderSummary, DueOrderSummaryAction } from '@homzhub/common/src/domain/models/DueOrderSummary';
import { Payment } from '@homzhub/common/src/domain/models/Payment';
import { IProcessPaymentPayload } from '@homzhub/common/src/modules/financials/interfaces';
import { ITypographyProps } from '@homzhub/common/src/components/atoms/Typography';
import {
  DuePaymentActions,
  IDueOrderSummaryAction,
  IPaymentParams,
  IPaymentPayload,
} from '@homzhub/common/src/domain/repositories/interfaces';
import { IListData } from '@homzhub/common/src/domain/models/Asset';
import { ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

const DuesOrderSummary = (): React.ReactElement | null => {
  const { goBack, navigate } = useNavigation();
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const currentDueId = useSelector(FinancialSelectors.getCurrentDueId);

  const [areCoinsApplied, setAreCoinsApplied] = useState(false);
  const [isPromoSelected, setIsPromoSelected] = useState(false);
  const [isPromoApplied, setIsPromoApplied] = useState(false);
  const [isPromoFailed, setIsPromoFailed] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [orderSummary, setOrderSummary] = useState<DueOrderSummary | null>(null);

  const resetStates = (): void => {
    setAreCoinsApplied(false);
    setIsPromoApplied(false);
    setIsPromoFailed(false);
    setIsPromoSelected(false);
  };

  useEffect(() => {
    getOrderSummary().then();
    return (): void => {
      resetStates();
    };
  }, []);

  const getOrderSummary = async (): Promise<void> => {
    try {
      setIsLoading(true);
      const response = await LedgerRepository.getDueOrderSummary(currentDueId);
      setOrderSummary(response);
      setIsLoading(false);
    } catch (e) {
      setIsLoading(false);
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details.message), statusCode: e.details.statusCode });
    }
  };

  const handleAction = async (payload: IDueOrderSummaryAction): Promise<void> => {
    try {
      setIsLoading(true);
      const response = await LedgerRepository.dueOrderSummaryAction(currentDueId, payload);
      setOrderSummary(response);
      if (isPromoSelected) {
        setIsPromoApplied(true);
      }
      if (isPromoFailed) {
        setIsPromoFailed(false);
      }
      setIsLoading(false);
    } catch (e) {
      setIsLoading(false);
      if (isPromoSelected) {
        setIsPromoFailed(true);
      }
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details.message), statusCode: e.details.statusCode });
    }
  };

  const handleCoins = async (): Promise<void> => {
    const payload: IDueOrderSummaryAction = {
      action: DueOrderSummaryAction.REFRESH_INVOICE_AMOUNT,
      coins: orderSummary?.coins?.currentBalance,
    };
    await handleAction(payload);
  };

  const handlePromo = async (code: string): Promise<void> => {
    const payload: IDueOrderSummaryAction = {
      action: DueOrderSummaryAction.REFRESH_INVOICE_AMOUNT,
      promo_code: code,
    };
    await handleAction(payload);
  };

  const onOrderPlaced = (paymentParams: IPaymentParams): void => {
    setIsLoading(true);
    const getBody = (): IPaymentPayload => {
      // Payment successful
      if (paymentParams.razorpay_payment_id) {
        const { razorpay_order_id, razorpay_payment_id, razorpay_signature } = paymentParams;
        return {
          action: DuePaymentActions.PAYMENT_CAPTURED,
          payload: { razorpay_order_id, razorpay_payment_id, razorpay_signature },
        };
      }
      // Payment cancelled
      const { razorpay_order_id } = paymentParams;
      return {
        action: DuePaymentActions.PAYMENT_CANCELLED,
        payload: { razorpay_order_id },
      };
    };

    const payload: IProcessPaymentPayload = {
      data: getBody(),
      onCallback: (status) => {
        if (status && paymentParams.razorpay_payment_id) {
          navigate(ScreensKeys.FinancialsLandingScreen);
          AlertHelper.success({
            message: t('assetFinancial:paidSuccessfully', { amount: orderSummary?.totalPriceFormatted }),
          });
        }
        setIsLoading(false);
      },
    };

    dispatch(FinancialActions.processPayment(payload));
  };

  const initiatePayment = async (): Promise<Payment | null> => {
    const payload: IDueOrderSummaryAction = {
      action: DueOrderSummaryAction.TRIGGER_PAYMENT,
      ...(isPromoApplied && { promo_code: orderSummary?.promo?.code }),
      ...(areCoinsApplied && { coins: orderSummary?.coins?.coinsUsed }),
    };
    try {
      const response = await LedgerRepository.dueOrderSummaryAction(currentDueId, payload);
      return response.order;
    } catch (e) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details.message), statusCode: e.details.statusCode });
      return null;
    }
  };

  const onToggleCoins = (): void => {
    // Unselecting the coins -> Call GET again
    if (areCoinsApplied) {
      getOrderSummary().then(() => {
        setAreCoinsApplied(false);
      });
      return;
    }
    if (orderSummary && orderSummary.coins && orderSummary.coins.currentBalance > 0) {
      handleCoins().then(() => {
        setAreCoinsApplied(true);
      });
    }
  };

  const onTogglePromoCode = (): void => {
    if (isPromoSelected) {
      if (isPromoApplied) {
        getOrderSummary().then();
        setIsPromoApplied(false);
      }
      setIsPromoFailed(false);
    }
    setIsPromoSelected((initValue) => !initValue);
  };

  const onApplyPromo = (code: string): void => {
    // Call API with applied promo.
    if (code.length) {
      handlePromo(code).then();
      return;
    }
    // For empty promo, refresh the data.
    onClearPromo();
  };

  const onClearPromo = (): void => {
    setIsPromoFailed(false);
    setIsPromoApplied(false);
    getOrderSummary().then();
  };

  const ShowInvoiceSummary = (): React.ReactElement | null => {
    if (!orderSummary) return null;

    const keyExtractor = (item: IListData, index: number): string => `${item.label}[${index}]`;

    const renderItem = ({ item }: { item: IListData }): React.ReactElement => (
      <View style={styles.itemsContainer}>
        <Text type="small" textType={item.isTitle ? 'semiBold' : 'regular'} style={styles.listValue}>
          {item.label}
        </Text>
        <Text type="small" textType={item.isTitle ? 'semiBold' : 'regular'} style={styles.listValue}>
          {item.value}
        </Text>
      </View>
    );

    const TotalView = (): React.ReactElement => {
      return (
        <View style={styles.totalView}>
          <Divider containerStyles={styles.divider} />
          <View style={styles.totalContent}>
            <Text type="small" textType="semiBold" style={styles.totalText}>
              Total
            </Text>
            <Text type="small" textType="semiBold" style={styles.totalText}>
              {orderSummary.totalBasePriceFormatted}
            </Text>
          </View>
        </View>
      );
    };

    return (
      <View style={styles.topSummaryContainer}>
        <FlatList data={orderSummary.userInvoiceItemsFormatted} keyExtractor={keyExtractor} renderItem={renderItem} />
        <TotalView />
      </View>
    );
  };

  if (!orderSummary) return <Loader visible />;

  const {
    asset,
    dueTitle,
    coins,
    dueOrderSummaryList,
    totalPricePayableFormatted,
    currency,
    invoiceTitle,
    countryFlag,
  } = orderSummary;

  const primaryAddressStyles: ITypographyProps = {
    variant: 'text',
    size: 'small',
  };

  const subAddressStyles: ITypographyProps = {
    size: 'regular',
    variant: 'label',
  };

  const onBackIconPress = (): void => {
    resetStates();
    goBack();
  };

  return (
    <Screen
      headerProps={{
        title: t('property:orderSummary'),
        onIconPress: onBackIconPress,
      }}
      contentContainerStyle={styles.screenContentContainer}
      scrollEnabled
      isLoading={isLoading}
    >
      <View style={styles.containerPadding}>
        <Text type="small" textType="regular" style={styles.orderSummaryText}>
          {t('property:orderSummary')}
        </Text>
        <PropertyAddressCountry
          primaryAddress={asset ? asset.projectName : dueTitle}
          primaryAddressTextStyles={primaryAddressStyles}
          subAddress={asset ? asset.formattedAddressWithCity : invoiceTitle}
          subAddressTextStyles={subAddressStyles}
          containerStyle={styles.addressContainer}
          countryFlag={countryFlag}
        />
        {orderSummary && <ShowInvoiceSummary />}
      </View>
      <HomzhubCoins
        onToggle={onToggleCoins}
        disabled={isPromoSelected || Boolean(coins && coins.currentBalance <= 0)}
        selected={areCoinsApplied}
        coins={coins ?? new Coins()}
        hasBalanceInNewLine
        containerStyle={styles.coinsContainer}
        showCoinCount={Boolean(coins && coins.currentBalance > 0)}
      />
      <View style={styles.containerPadding}>
        <PromoCode
          type="regular"
          onApplyPromo={onApplyPromo}
          isPromoApplied={isPromoApplied}
          isError={isPromoFailed}
          onClear={onClearPromo}
          hasToggleButton
          isToggleButtonSelected={isPromoSelected}
          onToggle={onTogglePromoCode}
          containerStyles={styles.promoContainer}
          isToggleButtonDisabled={areCoinsApplied}
        />
        <OrderSummary
          summaryList={dueOrderSummaryList}
          amountPayableText={totalPricePayableFormatted}
          currency={currency}
          containerStyle={styles.orderSummaryContainer}
          showOrderSummaryHeader={false}
          hasDottedDivider
        />
        {orderSummary.totalPricePayable > 0 && (
          <PaymentGateway
            type="primary"
            title={t('assetFinancial:payNow')}
            containerStyle={styles.payButton}
            // @ts-ignore
            initiatePayment={initiatePayment}
            paymentApi={onOrderPlaced}
          />
        )}
        <View style={styles.secureView}>
          <Icon name={icons.badge} color={theme.colors.darkTint7} size={28} />
          <Label type="large" textType="semiBold" style={styles.secureText}>
            {t('property:securePayment')}
          </Label>
        </View>
      </View>
    </Screen>
  );
};

export default DuesOrderSummary;

const styles = StyleSheet.create({
  listValue: {
    color: theme.colors.darkTint3,
  },
  totalText: {
    color: theme.colors.darkTint2,
  },
  totalView: {
    marginBottom: 16,
  },
  totalContent: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingTop: 16,
  },
  divider: {
    borderWidth: 1,
    borderColor: theme.colors.darkTint7,
  },
  itemsContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 16,
  },
  screenContentContainer: {
    paddingVertical: 20,
    paddingHorizontal: 0,
    backgroundColor: theme.colors.white,
  },
  addressContainer: {
    marginTop: 14,
    marginBottom: 10,
  },
  coinsContainer: {
    backgroundColor: theme.colors.moreSeparator,
    paddingHorizontal: 16,
  },
  orderSummaryContainer: {
    marginHorizontal: 0,
  },
  orderSummaryText: {
    color: theme.colors.darkTint4,
  },
  topSummaryContainer: {
    marginBottom: 10,
  },
  promoContainer: {
    marginHorizontal: 0,
  },
  payButton: {
    marginHorizontal: 16,
    marginTop: 12,
  },
  secureView: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: 10,
    justifyContent: 'center',
  },
  secureText: {
    color: theme.colors.darkTint7,
    marginLeft: 6,
  },
  containerPadding: {
    paddingHorizontal: 16,
  },
});
