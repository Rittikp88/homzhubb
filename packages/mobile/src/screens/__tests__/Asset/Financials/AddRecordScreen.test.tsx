// @ts-nocheck
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AddRecordScreen } from '@homzhub/mobile/src/screens/Asset/Financials/AddRecordScreen';

describe('Add Record Screen', () => {
  let component: ShallowWrapper;

  beforeEach(() => {
    component = shallow(<AddRecordScreen t={(key: string): string => key} />);
  });

  it('should render snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
