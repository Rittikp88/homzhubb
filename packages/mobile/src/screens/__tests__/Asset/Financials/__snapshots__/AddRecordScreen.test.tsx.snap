// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Record Screen should render snapshot 1`] = `
<Fragment>
  <Memo(AnimatedProfileHeader)
    title="financial"
  >
    <HeaderCard
      clear={0}
      onClearPress={[Function]}
      onIconPress={[Function]}
      renderItem={[Function]}
      subTitle="common:clear"
      title="addRecords"
    />
  </Memo(AnimatedProfileHeader)>
  <Loader
    visible={false}
  />
</Fragment>
`;
